Refactor TextFile.lines()

dtrebbien opened the pull request:
Introduces a LinesStream utility class that is responsible for closing
the underlying file handle.

Switches to calling Files.newBufferedReader() instead of wrapping the
InputStream returned by Files.newInputStream() in a BufferedReader of
an InputStreamReader.

Creates the file reader at the time that lines() is invoked, instead of
at the time that iterator() of the returned Iterable is invoked. This
can help prevent confusion as to the cause of an exception, whether it's
because of a problem opening the file or reading from it.

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* RFE: Developer: Introduce new `jenkins.util.io.LinesStream` class for iterating through files without fully caching them in memory

Maybe also: 

* RFE: Correctly propagate exceptions when reading Slave2Master security configuration file

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

oleg-nenashev commented on core/src/main/java/hudson/util/TextFile.java:
```diff
@@ -81,34 +146,8 @@ public String read() throws IOException {
-                }
-            }
-        };
+    public @Nonnull LinesStream lines() throws IOException {
```
The exception needs to be documented at least

dtrebbien replied:
Done

oleg-nenashev commented on core/src/main/java/hudson/util/TextFile.java:
```diff
@@ -50,6 +51,70 @@
+     * {@link #iterator()} to obtain a second or subsequent <code>Iterator</code>
+     * throws <code>IllegalStateException</code>.
+     */
+    public class LinesStream implements Closeable, Iterable<String> {
```
`@since TODO` or make it `@Restricted` if there is no plan to use it externally

oleg-nenashev replied:
Maybe you want to create a new utility class and pass file as a constructor parameter. There are other use-cases for such class, so no need to keep it nested

dtrebbien replied:
I extracted the class to top-level and added `@since TODO`.

oleg-nenashev commented on core/src/main/java/hudson/util/TextFile.java:
```diff
@@ -50,6 +51,70 @@
+        private transient Iterator<String> iterator;
+
+        public LinesStream() throws IOException {
+            in = Files.newBufferedReader(file.toPath()); // uses UTF-8 by default
```
🐛 It does not longer handle `InvalidPathException`. Use `hudson.Util.fileToPath()`

dtrebbien replied:
Done

oleg-nenashev reviewed (changes requested):
`InvalidPathException` is not handled anymore

oleg-nenashev commented:
@dtrebbien we do not get notifications when somebody does a force push. And generally there is no need to force push, we can squash commits during the merge if needed

dtrebbien commented:
Oh, sorry.

To summarize what I've done via force-pushing:

 * I addressed your changes from the first revision of the patch.
 * I relocated the new `LinesStream` utility class to the `jenkins.util.io` package.
 * I modified the documentation of TextFile.lines() to mention that the caller must close the returned `LinesStream`.

dtrebbien committed:
Refactor TextFile.lines()

Introduces a LinesStream utility class that is responsible for closing
the underlying file handle.

Switches to calling Files.newBufferedReader() instead of wrapping the
InputStream returned by Files.newInputStream() in a BufferedReader of
an InputStreamReader.

Creates the file reader at the time that lines() is invoked, instead of
at the time that iterator() of the returned Iterable is invoked. This
can help prevent confusion as to the cause of an exception, whether it's
because of a problem opening the file or reading from it.

dtrebbien commented:
Rebased to resolve conflict.

dtrebbien committed:
Merge remote-tracking branch 'origin/master' into refactor-TextFile-lines

dtrebbien committed:
Delete a duplicate import and a now-unused import

oleg-nenashev commented:
Sorry, I will review it eventually. JEP-200 and GSoC soak all my time this month

oleg-nenashev commented on core/src/main/java/jenkins/security/s2m/ConfigFile.java:
```diff
@@ -26,15 +27,17 @@ public ConfigFile(File file) {
     protected abstract COL readOnly(COL base);
 
-    public synchronized void load() {
+    public synchronized void load() throws IOException {
```
It breaks the source compatibility IIUC. I would rather introduce a new one and deprecate `load()`

daniel-beck replied:
Why do we care about source compatibility?

oleg-nenashev replied:
PCT?

oleg-nenashev commented on core/src/main/java/jenkins/security/s2m/ConfigFile.java:
```diff
@@ -79,8 +82,13 @@ public COL get() {
+                    try {
+                        load();
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
```
It would be rather preferable to create a new method with error propagation

oleg-nenashev commented on core/src/main/java/jenkins/util/io/LinesStream.java:
```diff
@@ -0,0 +1,114 @@
+                    }
+                    return r;
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
```
Should be at least documented in the iterator Javadoc

oleg-nenashev replied:
OTOH it does not make it worse being compared to the original code

oleg-nenashev commented on core/src/main/java/jenkins/util/io/LinesStream.java:
```diff
@@ -0,0 +1,114 @@
+ * @since TODO
+ */
+@CleanupObligation
+public class LinesStream implements Closeable, Iterable<String> {
```
Maybe `@Restricted(NoExternalUse.class)`? No strong opinion. Runtimized IOException on read error should be documented if it goes to public API

oleg-nenashev reviewed (approved):
@dtrebbien it looks good to me excepting one minor source compatibility thing. Would it be possible to address this comment so that I merge it?

dtrebbien committed:
Introduce load2()

dtrebbien committed:
Add a @Deprecated annotation

dtrebbien commented:
I have now added ConfigFile.load2() and deprecated the original ConfigFile.load() method.

oleg-nenashev commented:
Added the changelog entries

daniel-beck commented on core/src/main/java/jenkins/security/s2m/ConfigFile.java:
```diff
@@ -63,7 +91,7 @@ public synchronized void set(String newContent) throws IOException {
 
         write(newContent);
-        load();
+        load2();
```
Caused https://issues.jenkins.io/browse/JENKINS-67061
